### DISCUSSION TITLE
Populating drop-down options in FHIR Core from Resources in DB

### DISCUSSION THREAD

**Discussion by allan-on** [opensrp/fhircore#2467] [Ideas]
## Feature request
Syncing location (hierarchy) data from the server and populating drop-down options in a Questionnaire

## Context
In Questionnaires where the user needs to input location-related data such as the name of a physical location or Health Facility tied to a jurisdiction, a selection from a list would offer the best UX. 

### Server-side implementation
The implementation should provide syncing of location data from the server added through location hierarchy or practitioner assignments on FHIR Web.

- We can use the implementation that will be available on the [FHIR Gateway](https://github.com/opensrp/fhir-gateway) that allows getting Location and LocationHierarchy details as part of the Practitioner details 
- This will be once moving of the Practitioner details endpoint to plugins in the FHIR Gateway in https://github.com/opensrp/fhircore/issues/2213 is done

### Android FHIR SDK SDC Support
- Currently, it is only possible to use `answerExpressions` with an expression that takes a query and populates a drop-down item control. Support was added here https://github.com/google/android-fhir/issues/1039  

- This means that we can query against a Resource and use the values in a drop-down. This happens on loading of the Questionnaire and doesn't support multi-level selections. For instance, in a Questionnaire with multiple (multi-level) location selection options, a prior location selection within the same Questionnaire would not result in a change in the options in the next drop-down

- With `initialExpression` x-fhir-query support that will be introduced in https://github.com/google/android-fhir/issues/2036 it'll be possible to Query for and set an initially selected option

### FHIR Core implementation
- On FHIR Core we'd need to sync & persist the Location data using the location/practitioner details API.
  - We'll also need a  menu option to trigger syncing of locations/hierarchy manually whenever assignment updates are done on FHIR Web
  - There might be performance implications depending on the amount of data that is being synced. 
    -  Might be beneficial to be able to define which locations to sync in the hierarchy defined

a) Drop-down selection 
- The drop-down options can be populated using  `answerExpression`.  A sample implementation can be seen here https://github.com/google/android-fhir/issues/1830#issuecomment-1433133032

b) Multi-level selections
-  To implement multi-level selections that facilitate This would need SDK support for `answerExpression`s to work with an already loaded Questionnaire

----
#### Additional Context 
See [Loading of data on SDC widgets meeting notes](https://docs.google.com/document/d/130JS0nBU40zogZJhLV-BMqbBbi8NaTNgRJsV5_LbFsU/edit#)

**Comment by allan-on** [comment]
@f-odhiambo @dubdabasoduba @pld Please have a look at this and add any thoughts on the same

**Reply by pld** [reply]
This looks good, it's unclear to me that we would need any changes on the SDK or FHIRCore to support the syncing use-case you are describing here, I believe that's all supported through configurations now.

Then the code change needed here is to support answerExpression on the SDK. This is a P1 issue but looks quiet recently. For that do we need both https://github.com/google/android-fhir/issues/1830 and https://github.com/google/android-fhir/issues/2036?

**Reply by FikriMilano** [reply]
We need 1830 to initialize the location id value in the case we want to edit the response that has multi level location.
For context: when we use answer expression, the questionnaire item type is 'reference' a.k.a the id of the resource.

Not really sure why we need 2036.
@allan-on what do you think?

**Comment by FikriMilano** [comment]
> For instance, in a Questionnaire with multiple (multi-level) location selection options, a prior location selection within the same Questionnaire would not result in a change in the options in the next drop-down

Turns out, answer-expression can only access FHIRPath context from launch-context.
Therefore it cannot access:
- variables from variable expression
- other response item (simply)

@f-odhiambo remember our chat about _"can answerExpression works on rendered questionnaire? just like calculated-expression"_.
I have confirmed it yesterday and the answer is we can't, the reason is as I mentioned above.

The code:
https://github.com/google/android-fhir/blob/master/datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt#L585-L589

@pld @allan-on

**Reply by FikriMilano** [reply]
@f-odhiambo definitely need an SDK ticket for this, we can talk about it tomorrow if you need more explanation

**Reply by pld** [reply]
this feels like a silly question, but, can we use calculated expression here instead of answerExpression?

**Reply by FikriMilano** [reply]
not really, it's 2 different things.

calculated-expression
> This expression allows for dynamic calculation of answers to questions as other questions are answered

answer-expression
> Just like answerOption and answerValueSet, this is the third mechanism for capturing the allowed set of answers.

**Reply by pld** [reply]
can a calculated-expression generate a set of answers for another question, only a single answer? I guess no, that's what answer-expression is for.

But given the similarity in use-case, it sounds like answer-expression should function very similarly (or even the same as?) calculated-expression

**Reply by FikriMilano** [reply]
now you mention it, well it's very similar.

the only difference is:
calculated-expression: gives the answer selections
answer-expression: gives the answer options

**Comment by FikriMilano** [comment]
@pld @f-odhiambo @dubdabasoduba 
tracked by this sdk ticket https://github.com/google/android-fhir/issues/2052
I think someone from the google side will pick that up as Jing mentioned.
@MJ1998

**Comment by FikriMilano** [comment]
@pld @f-odhiambo @dubdabasoduba
I think we are unblocked, as long we create a snapshot w this branch
https://github.com/google/android-fhir/pull/2066

[device-2023-07-11-191923.webm](https://github.com/opensrp/fhircore/assets/62053304/82601376-53f2-4562-a1cb-aee40cb10fe2)

**Comment by SebaMutuku** [comment]
@FikriMilano when you run the app with locations, you need to click on the questionnaire  fields twice for it to load filtered resources from the local db. Have you experienced this?

**Reply by FikriMilano** [reply]
@SebaMutuku  can you share the steps to replicate that? including the branch, appId, and user credentials to login

**Reply by FikriMilano** [reply]
if it's using the questionnaire I shared, it doesn't need to double tap to the field.
https://drive.google.com/file/d/1aweK2JZ-3dzG9N7M2KoPwy_80sqZF9my/view

but you can try it, to make sure, by following these steps, and the same branch (use jajoo's [fork/branch](https://github.com/MJ1998/android-fhir/tree/2052)):
https://github.com/google/android-fhir/issues/2052

**Comment by SebaMutuku** [comment]
This is how it appears the first time
![image](https://github.com/opensrp/fhircore/assets/36365043/ec455221-db8f-4059-b36b-07d0e3adf7b1)

However, when you press the dash and press the drop-down again,  the` x-fhir-query` is executed. 
![image](https://github.com/opensrp/fhircore/assets/36365043/e0b43c53-56a4-4f66-96b0-c7a743fa64cb)

**Comment by SebaMutuku** [comment]
@FikriMilano can you try testing the drop-down on the questionnaire without this part  
 ```
  "_text": {
        "extension": [
          {
            "url": "http://hl7.org/fhir/StructureDefinition/cqf-expression",
            "valueExpression": {
              "language": "text/fhirpath",
              "expression": "'Find Patient From: ' + %city + ' City'"
            }
          }
        ]
      },
```

**Reply by FikriMilano** [reply]
@SebaMutuku are you referring to this questionnaire? (the one I wrote)
https://drive.google.com/file/d/1aweK2JZ-3dzG9N7M2KoPwy_80sqZF9my/view

or is it the one you wrote for locations?

**Reply by SebaMutuku** [reply]
@FikriMilano , the one you wrote. I notice if you remove the _text block, the questionnaire needs you to press on the fields twice for it to fetch data using `x-fhir-query`. Sometimes it's not. Let me know  when we can sync up on this

**Reply by FikriMilano** [reply]
@SebaMutuku  let me try again today, investigate a bit, and get back to you

**Reply by FikriMilano** [reply]
@SebaMutuku you're correct, it's oddly not working without the cqf expression (_text block), I had a quick look, haven't found the reason.

can you submit a bug report to the SDK?

**Comment by sharon2719** [comment]
@pld @allan-on Can I pick up on documenting this implementation? It will be very useful for the entire team to get to understand how this is done.

**Comment by pld** [comment]
Yes please go aheadOn Dec 18, 2023, at 02:00, Sharon Akinyi ***@***.***> wrote:﻿
@pld @allan-on Can I pick up on documenting this implementation? It will be very useful for the entire team to get to understand how this is done.
